src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,284 @@
+    {
+        protected abstract string CodeActionTitle { get; }
+
+        protected abstract bool CanRefactorToContainBlockStatements(SyntaxNode parent);
+        protected abstract SyntaxList<TStatementSyntax> GetStatements(SyntaxNode parent);
+        protected abstract SyntaxNode WithStatements(SyntaxNode parent, SyntaxList<TStatementSyntax> statements);
+
+        protected abstract TStatementSyntax CreateUsingStatement(TLocalDeclarationSyntax declarationStatement, SyntaxTriviaList sameLineTrivia, SyntaxList<TStatementSyntax> statementsToSurround);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+
+            var (declarationSyntax, _) =
+                await FindDisposableLocalDeclaration(document, span, context.CancellationToken).ConfigureAwait(false);
+
+            if (declarationSyntax != null)
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    CodeActionTitle,
+                    cancellationToken => IntroduceUsingStatementAsync(document, span, cancellationToken)));
+            }
+        }
+
+        private async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax =
+                root.FindNode(selection)?.GetAncestorOrThis<TLocalDeclarationSyntax>()
+                ?? root.FindTokenOnLeftOfPosition(selection.End).GetAncestor<TLocalDeclarationSyntax>();
+
+            if (declarationSyntax is null || !CanRefactorToContainBlockStatements(declarationSyntax.Parent))
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var disposableType = semanticModel.Compilation.GetSpecialType(SpecialType.System_IDisposable);
+            if (disposableType is null)
+            {
+                return default;
+            }
+
+            var operation = semanticModel.GetOperation(declarationSyntax, cancellationToken) as IVariableDeclarationGroupOperation;
+            if (operation?.Declarations.Length != 1)
+            {
+                return default;
+            }
+
+            var localDeclaration = operation.Declarations[0];
+            if (localDeclaration.Declarators.Length != 1)
+            {
+                return default;
+            }
+
+            var declarator = localDeclaration.Declarators[0];
+
+            var localType = declarator.Symbol?.Type;
+            if (localType is null)
+            {
+                return default;
+            }
+
+            var initializer = (localDeclaration.Initializer ?? declarator.Initializer).Value;
```

Comment:
Is there a case where both `localDeclaration` and `declarator` have null `Initializer`s or am I paranoid?

Reply:
I don't know, but paranoid sounds good to me until Roslyn is able to use nullable reference types! :+1: